clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumerImpl.java

```
@@ -1017,8 +1017,12 @@ private void maybeThrowInvalidGroupIdException() {
      * <p>
      * If the acknowledgement commit callback throws an exception, this method will throw an exception.
      */
-    private void handleCompletedAcknowledgements() {
-        processBackgroundEvents();
+    private void handleCompletedAcknowledgements(boolean onClose) {
+        // If the user gets any fatal errors, they will get these exceptions in the background queue.
+        // While closing, we ignore these exceptions so that the consumers close successfully.
+        processBackgroundEvents(onClose ? e -> (e instanceof GroupAuthorizationException
```

This is an extremely specific list of exceptions to ignore. I wonder why precisely these three make sense.

---

So I was trying to follow the pattern here when we ignore certain exceptions while unsubscribing - https://github.com/apache/kafka/blob/9a6e4f1dcc67df825b209defe0b3d09b8e2d8ff6/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumerImpl.java#L940.
We propagate fatal errors and metadata errors to the application thread, probably that's why these 3 exceptions were used. I will check once if there are any other exceptions which could come, we should probably ignore those as well.

---

Sharing some context in case it helps (related to handling this for the consumer `unsubscribe`/`close`)

In that case it made sense to ignore the `TopicAuth` and `GroupAuth` exceptions because those are the ones we could get in the HB to leave the group that we seemed to be irrelevant to the user wanting to leave already. 

We only considered errors that could come in the HB response, without having to worry about Metadata errors (like InvalidTopic) because, for the consumer case, the unsubscribe operation in not aware of those (metadata errors are not propagated via the generic error queue for the AsyncConsumer, so they won't come out when `processingBackgroundEvents` on unsubscribe or close)

But the ShareConsumer does get Metadata errors via the generic error queue, so I would expect that it makes sense to ignore the `InvalidTopic` in this case too (along with any other error that could come in a MetadataResponse). Maybe worth looking at the `handleTopicMetadata` in `KafkaApis` in case another error generated there seems sensible to ignore in this case (I don't see any other from a quick look but could be missing something)

---

Thanks @lianetm for the context here, this really helps.
Yes I did take a look at `handleTopicMetadata` in `KafkaApis`, I could not find any topic/auth related errors that we are missing. There could still be retriable errors received in the metadata response but those are not propagated to the application thread, so we need not add them here.
The current checks seem to cover all the cases then.